### PR TITLE
Add clustering features

### PR DESCRIPTION
**Pull request by lliiun-z**
Test plan - (Please fill in how you tested your changes)

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines. Don't forget to follow our [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution) for any code copied from other projects.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by highker** [comment]
Reference #15912 here as the initial comments are there.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java [hunk 1]
```diff
@@ -171,6 +179,13 @@ public HivePageSink(
                     .mapToInt(dataColumnNameToIdMap::get)
                     .toArray();
             BucketFunctionType bucketFunctionType = bucketProperty.get().getBucketFunctionType();
+
+            // TODO: Add correct switch for HIVE_COMPATILE and HIVE_CLUSTERING.
+            if (bucketProperty.get().getDistribution().get().isEmpty() && bucketProperty.get().getBucketFunctionType() == BucketFunctionType.HIVE_CLUSTERING)
+            {
+                bucketFunctionType = BucketFunctionType.HIVE_COMPATIBLE;
+            }
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java:187]: We probably don't need this once we use `CLUSTERED_BY`?

File: presto-common/src/main/java/com/facebook/presto/common/clustering/Distribution.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.facebook.presto.common.type.Type;
+import javafx.util.Pair;
+
+import java.util.ArrayList;
+import java.util.List;
+
+// Assume: We have 100 values to represent 100 percentiles
+// for the column for now.
+public final class Distribution
```
Comment (highker) [inline presto-common/src/main/java/com/facebook/presto/common/clustering/Distribution.java:11]: Let's try to reuse existing helpers like `Range` or `Block`. I think in this case, we could have a simple Block to represent values in sorted order.

**Comment by highker** [comment]
Seems the branch is broken. Could you squash them or clean it up? I once saw this branch structure from one of @tanjialiang's PR. Probably it is similar in this case.

File: presto-common/src/main/java/com/facebook/presto/common/clustering/Distribution.java (hunk 2 above)
Comment (lliiun-z) [reply presto-common/src/main/java/com/facebook/presto/common/clustering/Distribution.java:11]: Yeah, after we finalize the logic, I will refactor these classes.

**Comment by lliiun-z** [comment]
Close this one since we already has a new one.

**Pull request closed without merging**